flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskTierConsumerAgent.java

```diff
@@ -62,20 +74,36 @@ public void start() {
         // noop
     }
 
+    @Override
+    public void registerAvailabilityAndPriorityNotifier(AvailabilityAndPriorityNotifier notifier) {
+        this.notifier = notifier;
+    }
+
     @Override
     public Optional<Buffer> getNextBuffer(
             TieredStoragePartitionId partitionId,
             TieredStorageSubpartitionId subpartitionId,
             int segmentId) {
+        Optional<Buffer> buffer = Optional.empty();
+        Tuple2<CompletableFuture<NettyConnectionReader>, Integer> readerAndBufferIndex =
+                nettyConnectionReaders.get(partitionId).get(subpartitionId);
         try {
-            return nettyConnectionReaders
-                    .get(partitionId)
-                    .get(subpartitionId)
-                    .get()
-                    .readBuffer(segmentId);
+            buffer = readerAndBufferIndex.f0.get().readBuffer(segmentId);
         } catch (InterruptedException | ExecutionException e) {
-            throw new RuntimeException("Failed to get next buffer.", e);
+            ExceptionUtils.rethrow(e, "Failed to read buffer from memory tier.");
         }
+        buffer.ifPresent(
+                value -> {
+                    boolean isPriority = value.getDataType().hasPriority();
+                    checkNotNull(notifier)
+                            .notifyAvailableAndPriority(
+                                    partitionId,
+                                    subpartitionId,
+                                    isPriority,
+                                    isPriority ? readerAndBufferIndex.f1 : null);
+                    readerAndBufferIndex.f1 += 1;
```

Comment:
How the three tiers use the notifier seems quite similar. I wonder why this is inside each tier? Is it simply because the consumer client doesn't know the buffer index? If true, we may consider to include buffer index as part of the return value.